File: contracts/dre-contract/src/actions/evolve.ts
import { ContractState, ContractInput, ContractResult } from '../types'

export default function evolve(state: ContractState, input: ContractInput): ContractResult {
    if (!input.value) throw new ContractError('invalid arguments')

    const caller = SmartContract.transaction.owner

    if (caller != state.owner) throw new ContractError('only the owner can evolve')

    if (!state.canEvolve) throw new ContractError('cannot evolve contract')

    state.evolve = input.value

    return { state }
}

File: contracts/client-contract/src/actions/evolve.rs
use crate::types::{ ContractState, Evolve };
use warp_contracts::handler_result::WriteResult;

pub fn evolve(mut state: ContractState, input: Evolve) -> WriteResult<ContractState, String> {
    if !state.can_evolve || !state.is_direct_owner() {
        return WriteResult::ContractError("invalid owner".to_string())
    }
    
    state.evolve = Some(input.value);
    
    WriteResult::Success(state)
}